### DISCUSSION TITLE
Curtain motor script help needed (TUYA MCU)

### DISCUSSION THREAD

**Discussion by tonysprenk** [RavenSystem/esp-homekit-devices#2310] [MEPLHAA Scripts]
Hello all, 

I have curtain motors that communicate and are controlled with UART commands. I have created the following script to set the state of the curtain in the home app and to ensure that when the curtains are opened manually the stars is updated in HomeKit accordingly. However that last part is not working. Anything I might be missing? 
```
{
    "c": {
        "o": 8,                          // Enable OTA logging
        "r": [{
            "n": 12,                     // UART GPIO pin
            "s": 9600,                   // Baud rate
            "l": [15,15]                 // RX/TX buffer sizes
        }]
    },
    "a": [
        {   // Main Window Covering Service
            "t": 45,                     // Window Covering type
            "o": 19,                     // Opening time (seconds)
            "c": 19,                     // Closing time (seconds)
            "m": 0,                      // No position margin
            
            // Control Commands
            "0": {                       // Stop command
                "u": [{
                    "v": "55aa00060005660400010075"
                }]
            },
            "1": {                       // Open command
                "u": [{
                    "v": "55aa00060005660400010176"
                }]
            },
            "2": {                       // Close command
                "u": [{
                    "v": "55aa00060005660400010277"
                }]
            }
        },
        {   // Free Monitor for Manual Open Detection
            "t": 80,                     // Free Monitor type
            "h": 0,                      // No header processing
            "n": 24,                     // Message length
            "ff": -1,                    // No fixed field
            "fo": 100,                   // Field offset
            "pt": [                      // Pattern to match
                ["55aa00070008650200040000000379"]
            ],
            "dt": [1,0],                 // Data target
            "y0": [{                     // Action when pattern matched
                "0": {
                    "c": [[0,1,1,1]]     // Trigger open action
                },
                "v": 0
            }]
        },
        {   // Free Monitor for Manual Close Detection
            "t": 80,                     // Free Monitor type
            "h": 0,                      // No header processing
            "n": 24,                     // Message length
            "ff": -1,                    // No fixed field
            "fo": 100,                   // Field offset
            "pt": [                      // Pattern to match
                ["55aa0007000865020004000000037c"]
            ],
            "dt": [1,0],                 // Data target
            "y0": [{                     // Action when pattern matched
                "0": {
                    "c": [[0,0,1,1]]     // Trigger close action
                },
                "v": 0
            }]
        }
    ]
}
```

**Comment by RavenSystem** [comment]
Reverse engineering using original firmware is needed.

This document may help: 
[Guide-to-Interworking-with-the-Tuya-MCU.pdf](https://github.com/user-attachments/files/17799011/Guide-to-Interworking-with-the-Tuya-MCU.pdf)

**Comment by xrust83** [comment]
https://youtu.be/wSopCDhy8rs?si=dsrNfji-9neBnN83

the language of codes is universal. I think my experience will help you understand the analogy

https://youtu.be/sGAP4qEoSWs?si=zb3rF47AtYMMVjsZ

https://youtu.be/PeY7vvexAGA?si=iR3pjrK_hpmwxr8B

as you can see, Tuya did not particularly bother with different types of motors. your curtains or the team's blinds are the same

**Reply by xrust83** [reply]
Quickly reviewed what you analyze with a free monitor. to catch the value of curtains in percentage, you need to discard not the whole command, but part of it before posting the percentage bits

**Reply by tonysprenk** [reply]
Thank you! this is a great resource. However I think the commands is not the main issue but actually how I can implement it in the script. Any help on that would be greatly appreciated.

**Reply by xrust83** [reply]
the second video explains how to get a response from the blind from % position. the received answer must be recorded in current and target setting so that the curtain exits the closing or opening status.

**Reply by xrust83** [reply]
I don't understand what stars in homekit you are writing about, but you can send percentages of curtains as a characteristic to another service.

**Reply by tonysprenk** [reply]
Thank you my friend! It is working as expected! Two questions: 1. is there a way to ensure that the opening and closing time in the script is respected? I understand that with this setup when the curtain stops by itself because it can’t move any further the status is updated in home app. In most cases this works fine but I have 1 curtain that needs to stop before it cannot move any further because otherwise it will hit the radiator. 2. Is it possible to somehow implement the instruction detected command in the script?

**Reply by tonysprenk** [reply]
Also: do you have a schematic of the usb dongle? I want to recreate it with temperature and air quality sensors.

**Reply by xrust83** [reply]
Watch 3 videos - with roller blinds. there are commands where limits are set for the upper and lower position of the roller blind. I am not sure that it will work for ordinary curtains, but during self-calibration, the curtain somehow remembers the position of 100-0% and, after calibration, approaches these points carefully without pulling the belt.

**Reply by xrust83** [reply]
I use Sprint Layout. You can find the board file on google disc

**Reply by tonysprenk** [reply]
It's possible for you to export the schematic to another format? I don't have a PC to use spring t layout. only Mac.
